crates/red_knot_python_semantic/src/types/context.rs

```diff
@@ -131,6 +181,17 @@ impl fmt::Debug for InferContext<'_> {
     }
 }
 
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Default)]
+pub(crate) enum InNoTypeCheck {
+    /// The inference might be in a `no_type_check` block but only if any
+    /// enclosing function is decorated with `@no_type_check`.
+    #[default]
+    IfEnclosingFunction,
```

Comment:
The name of this variant is quite verbose but doesn't really add much clarity IMO (it wasn't at all obvious at the usage sites what this variant represented; I still had to jump to the enum definition and read the doc comment to understand what it meant). I'd probably just call it `Maybe` or `Possibly`. The doc comment is great!